### ISSUE TITLE
uni export: use of undefined constant '%const.0'. llc bug? Methodology bug?

### ISSUE THREAD

**Issue by matsc-at-sics-se** [matsc-at-sics-se/unison#26]
Reproducer:
[kff-clang-fft.ll.txt](https://github.com/matsc-at-sics-se/unison/files/1971686/kff-clang-fft.ll.txt)

```
$ llc kff-clang-fft.ll -unison -unison-verbose -unison-single-function=mk_twiddles -O2 -disable-post-ra -disable-tail-duplicate -disable-branch-fold -disable-block-placement -disable-phi-elim-edge-splitting -disable-hexagon-cfgopt
[...]
/usr/local/bin/uni: uni import --target=X86 /tmp/unison-989a76.mir -o /tmp/unison-989a76.uni --function=mk_twiddles --maxblocksize=25 --goal=speed
/usr/local/bin/uni: uni linearize --target=X86 /tmp/unison-989a76.uni -o /tmp/unison-989a76.lssa.uni
/usr/local/bin/uni: uni extend --target=X86 /tmp/unison-989a76.lssa.uni -o /tmp/unison-989a76.ext.uni
/usr/local/bin/uni: uni augment --target=X86 /tmp/unison-989a76.ext.uni -o /tmp/unison-989a76.alt.uni
/usr/local/bin/uni: uni normalize --target=X86 /tmp/unison-989a76.asm.mir -o /tmp/unison-989a76.llvm.mir
/usr/local/bin/uni: uni model --target=X86 /tmp/unison-989a76.alt.uni -o /tmp/unison-989a76.json --basefile=/tmp/unison-989a76.llvm.mir +RTS -K20M -RTS
/usr/local/bin/gecode-presolver: gecode-presolver -o /tmp/unison-989a76.ext.json --verbose /tmp/unison-989a76.json
/usr/local/bin/gecode-solver: gecode-solver -o /tmp/unison-989a76.out.json --verbose /tmp/unison-989a76.ext.json
/usr/local/bin/uni: uni export --target=X86 /tmp/unison-989a76.alt.uni -o /tmp/unison-989a76.unison.mir --basefile=/tmp/unison-989a76.llvm.mir --solfile=/tmp/unison-989a76.out.json
error: /tmp/unison-989a76.unison.mir:2075:41: use of undefined constant '%const.0'
    %xmm0 = VMULSDrm %xmm0, %rip, 1, _, %const.0, _
                                        ^
```

Note `-unison-single-function=mk_twiddles`, because the *.ll contains other functions that I want to ignore for the time being.  *.mir, *.asm.mir, *.llvm.mir, *.unison.mir all mention `const.0` but do not define it.

Is there a better way to compile a single function out of a *.ll?

**Comment by matsc-at-sics-se** [comment]
The problem prevails on a smaller input file if I omit the `-unison-single-function` flag.

**Comment by robcasloz** [comment]
Hi, the problem is that Unison drops the information about constants stored in memory (constant pool) that is present in the input .mir file:

```
constants:       
  - id:              0
    value:           double 0x400921FB54442D18
    alignment:       8
  - id:              1
    value:           '<4 x i32> <i32 -2147483648, i32 -2147483648, i32 -2147483648, i32 -2147483648>'
    alignment:       16 
```

This information should be kept and passed through all the way to the .unison.mir file. For that, we need to parse it in `src/unison/src/MachineIR/Parser.hs` (easy because it is in YAML format, we do this already e.g. to parse stack frame information), create data structures to store it in the MIR and Unison IR representations, and emit it in `src/unison/src/MachineIR/Instances.hs`. This task is not very challenging but can be a bit tedious.

**Comment by matsc-at-sics-se** [comment]
The problem prevents `llc -unison` from generating *.s for the AVX case study:
```
/usr/local/bin/uni: uni export --target=X86 /tmp/unison-17b4ae.alt.uni -o /tmp/unison-17b4ae.unison.mir --basefile=/tmp/unison-17b4ae.llvm.mir --solfile=/tmp/unison-17b4ae.out.json
error: /tmp/unison-17b4ae.unison.mir:1268:34: use of undefined constant '%const.0'
    %xmm0 = VMOVSSrm %rip, 1, _, %const.0, _
                                 ^
```

**Comment by robcasloz** [comment]
Addressed by commit 3b89f90bbaa.
